src/Altinn.App.Core/Internal/Process/ProcessEventDispatcher.cs

```diff
@@ -0,0 +1,141 @@
+using Altinn.App.Core.Configuration;
+using Altinn.App.Core.Interface;
+using Altinn.App.Core.Internal.Process.Elements;
+using Altinn.Platform.Storage.Interface.Enums;
+using Altinn.Platform.Storage.Interface.Models;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+
+namespace Altinn.App.Core.Internal.Process;
+
+/// <summary>
+/// Default implementation of the process event dispatcher
+/// </summary>
+class ProcessEventDispatcher : IProcessEventDispatcher
+{
+    private readonly IInstance _instanceService;
+    private readonly IProcess _processClient;
+    private readonly ITaskEvents _taskEvents;
+    private readonly IAppEvents _appEvents;
+    private readonly IEvents _eventsService;
+    private readonly bool _registerWithEventSystem;
+    private readonly ILogger<ProcessEventDispatcher> _logger;
+
+    public ProcessEventDispatcher(
+        IInstance instanceService, 
+        IProcess processClient, 
+        ITaskEvents taskEvents, 
+        IAppEvents appEvents, 
+        IEvents eventsService, 
+        IOptions<AppSettings> appSettings,
+        ILogger<ProcessEventDispatcher> logger)
+    {
+        _instanceService = instanceService;
+        _processClient = processClient;
+        _taskEvents = taskEvents;
+        _appEvents = appEvents;
+        _eventsService = eventsService;
+        _registerWithEventSystem = appSettings.Value.RegisterEventsWithEventsComponent;
+        _logger = logger;
+    }
+
+    /// <inheritdoc/>
+    public async Task<Instance> UpdateProcessAndDispatchEvents(Instance instance, Dictionary<string, string>? prefill, List<InstanceEvent>? events)
+    {
+        await HandleProcessChanges(instance, events, prefill);
+
+        // need to update the instance process and then the instance in case appbase has changed it, e.g. endEvent sets status.archived
+        Instance updatedInstance = await _instanceService.UpdateProcess(instance);
```

Comment:
Should this class, which is responsible for dispatching events, update and save the process state on the instance? Seems like a side effect.